amethyst_audio/src/systems/dj.rs

```diff
@@ -3,41 +3,79 @@ use log::error;
 use thread_profiler::profile_scope;
 
 use amethyst_assets::AssetStorage;
-use amethyst_core::ecs::prelude::*;
+use amethyst_core::ecs::*;
+use amethyst_error::Error;
 
 use crate::{
     output::init_output,
     sink::AudioSink,
     source::{Source, SourceHandle},
 };
 
-/// Calls a closure if the `AudioSink` is empty.
-pub fn build_dj_system<F, R>(
-    mut f: F,
-) -> Box<dyn FnOnce(&mut World, &mut Resources) -> Box<dyn Schedulable>>
+/// Dj system bundle which is the default way to construct dj system as it initializes any required resources.
+#[derive(Debug)]
+pub struct DjSystemBundle<F, R>
+where
+    F: FnMut(&mut R) -> Option<SourceHandle> + Send + Sync + 'static,
+    R: Send + Sync + 'static,
+{
+    f: Option<F>,
+    _phantom: std::marker::PhantomData<R>,
+}
+
+impl<F, R> DjSystemBundle<F, R>
+where
+    F: FnMut(&mut R) -> Option<SourceHandle> + Send + Sync + 'static,
+    R: Send + Sync + 'static,
+{
+    /// Creates a new [DjSystemBundle] where [f] is a function which produces music [SourceHandle].
+    pub fn new(f: F) -> Self {
+        Self {
+            f: Some(f),
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<F, R> SystemBundle for DjSystemBundle<F, R>
 where
     F: FnMut(&mut R) -> Option<SourceHandle> + Send + Sync + 'static,
     R: Send + Sync + 'static,
 {
-    Box::new(|_world: &mut World, resources: &mut Resources| {
+    fn load(
+        &mut self,
+        _world: &mut World,
+        resources: &mut Resources,
+        builder: &mut DispatcherBuilder,
+    ) -> Result<(), Error> {
         init_output(resources);
-        SystemBuilder::<()>::new("DjSystem")
-            .read_resource::<AssetStorage<Source>>()
-            .read_resource::<Option<AudioSink>>()
-            .write_resource::<R>()
-            .build(move |_commands, _world, (storage, sink, res), _queries| {
-                #[cfg(feature = "profiler")]
-                profile_scope!("dj_system");
+        builder.add_system(build_dj_system(self.f.take().unwrap()));
```

Comment:
> Wow, very impressive work!
> I have a few remarks, but nothing major.
> Also, I think all the dispatcher stuff should be moved into a separate crate. legion_dispatch maybe?
> Thanks a lot! ♥️

Thanks!

Dispatcher is just a single file so I don't know if it's worth moving it into a crate?

All the `EventChannel<Event>` unwraps should not fail if you use `Application`, because it is inserted before anything else. It is possible to change it to `get_mut_or_default`, but then it would silently wont work if you forgot to insert your custom input system. Maybe I could just add `expect` everywhere.